closes #1 Add OAuth2 methods to allow integration with OAuth2 clients

billoley opened the pull request:
closes #1 Add OAuth2 methods to allow integration with OAuth2 clients

These operations will support the OAuth 2.0 Authorization Code Grant described here:
https://tools.ietf.org/html/rfc6749

Added some comments to aid in the review.

billoley committed:
closes #1 Add OAuth2 methods to allow integration with OAuth2 clients

work

more work

Implement oauth2 service

work

more work

closes #1 Add OAuth2 methods to allow integration with OAuth2 clients

closes #1 Add OAuth2 methods to allow integration with OAuth2 clients

closes #1 Add OAuth2 methods to allow integration with OAuth2 clients

closes #1 Add OAuth2 methods to allow integration with OAuth2 clients

closes #1 Add OAuth2 methods to allow integration with OAuth2 clients

billoley commented on api/src/main/java/datawave/security/authorization/JWTTokenHandler.java:
```diff
@@ -93,32 +95,40 @@ public JWTTokenHandler(Certificate cert, Key signingKey, long jwtTtl, TimeUnit j
                 .setExpiration(expirationDate)
-                .claim(PRINCIPALS_CLAIM, users)
+                .claim(claimName, users)
                 .signWith(SignatureAlgorithm.RS512, signingKey)
```
GZIP is not specified by the applicable standard, only DEFLATE (ZLIB)

billoley commented on api/src/main/java/datawave/security/authorization/DatawaveUser.java:
```diff
@@ -37,17 +39,25 @@
     
```
Backward compatible constructor that excludes the email parameter.  If the other Datawave repos that use this form get refactored, then this can be removed.

billoley commented on api/src/main/java/datawave/security/authorization/JWTTokenHandler.java:
```diff
@@ -133,8 +143,12 @@ private CustomJWTBuilder(ObjectMapper objectMapper) {
         }
         
```
jjwt signature of this method changed wrt Exception thrown

billoley commented on service/src/test/java/datawave/microservice/authorization/AuthorizationServiceTest.java:
```diff
@@ -44,9 +46,9 @@
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.fail;
 
```
Use @ActiveProfiles and @Profile to limit when test-only classes get instantiated (and injected)

keith-ratcliffe commented on service/src/main/java/datawave/microservice/authorization/OAuthOperations.java:
```diff
@@ -0,0 +1,203 @@
+        }
+        
+        Collection<SubjectIssuerDNPair> userDnsToLookupAndAdd = new LinkedHashSet<>();
+        if (grant_type.equals("authorization_code")) {
```
Maybe promote the grant_type values to class-scoped constants? Assuming they're not already defined publicly somewhere else

keith-ratcliffe commented on service/src/main/java/datawave/microservice/authorization/OAuthOperations.java:
```diff
@@ -0,0 +1,203 @@
+        } else {
+            Caffeine<Object,Object> caffeine = Caffeine.newBuilder();
+            caffeine.expireAfterWrite(authCodeTtl, TimeUnit.SECONDS);
+            CACHE = caffeine.build();
```
Is local cache sufficient for this scenario? Wondering if something backed by `@CacheConfig` / `@Cacheable` could be needed here, similar to  our CachedDatawaveUserService impls, for the case when multiple instances of the service are deployed and we need to plug in something like Hazelcast, etc

keith-ratcliffe reviewed (dismissed):
Would also be good to document the new API on the README.md

billoley committed:
Use autowired cache instead of local map for authorization requests, use constants for some parameters

Use autowired cache instead of local map for authorization requests plus use constants for some parameters

Use constants

billoley committed:
Add ehcache resource for testing, documentation
